compress-mpq.c

```diff
@@ -668,13 +643,13 @@ void PackFiles(void *arguments){
         uint32_t flags;
         int foundCache = 0;
 
-        if (globals.useCache) {
-            foundCache = ReadCache(*path, insize, (unsigned char*)content, out, &outsize, &flags);
+        if(globals.useCache) {
+            foundCache = ReadCache(*path, (const size_t)insize, (const unsigned char*)content, out, &outsize, &flags);
         }
-        if (foundCache == 0){
+        if(foundCache == 0){
             PackFile((unsigned char*)content, insize, insize + sotSize, out, &outsize, &flags);
             if (globals.useCache){
-                CachePacked(*path, insize, outsize, flags, (unsigned char*)content, out);
+                CachePacked(*path, out, outsize, content, insize);
```

Comment:
Let's assume we only use the hash as the filename. Then when packing two files in parallel which happen to have the same content, this could lead to race conditions. Maybe grab a lock here.

Reply:
Maybe also for writing.

Reply:
Yet another reason for including ``in_archive_path``!

An alternative scenario entailing similar risks would be for someone to run more N>1 instances of ``compress-mpq`` at the same time. But that doesn't make sense, when they could reach the same result by running sequentially ``compress-mpq`` N times with N times as many threads, and even more if ``compress-mpq`` were to be arbitrarily run on untrusted users requests.

Despite all this, I guess I can just add a ``.lockfile`` check for all cache operations?

Reply:
> An alternative scenario entailing similar risks would be for someone to run more N>1 instances of `compress-mpq` at the same time. But that doesn't make sense, when they could reach the same result by running sequentially `compress-mpq` N times with N times as many threads, and even more if `compress-mpq` were to be arbitrarily run on untrusted users requests.

I agree.

> Despite all this, I guess I can just add a `.lockfile` check for all cache operations?

Maybe a better option, also in regards to the likes of shutdown, would be to write the cache file at first in a temporary file at first and then rename them, which is considered atomic. This would also remoe the need for a lockfile?

Reply:
Conceptually, that's ok, but Windows renaming in pure C isn't any fun :cry:

Reply:
Yeah i don't want to bikeshed too much but temporary files seem quite clean to me. 🤷 
Also personally i compile with mingw anyways so i don't care too much about winapi.
I will look over the code in the following days.